Show Blog Posts from API

clementinelove opened the pull request:
# UPDATED 10 Sep 2024

- Now API client is updated to use the new response format.
- Static post related function calls removed
- Currently the blog page only shows posts from API server, notice that old markdown post are NOT added to server yet!

---

## Last Week Update

Remote Blogs fetched from API are now presented along with static blogs, screenshots are attached below.

## Important Changes: 
- Remote blog `slug` property is not very useful: because we need blog ID to fetch blog content, the URL slug part must contain the blog ID i.e. something like `/blog/1` rather than  `/blog/post/blog-1.md`. One drawback is that these articles won't have a descriptive url, but I guess this is pretty much unavoidable unless our authors always give their articles a unique url path name.
- Since these remote blogs having integer ID, I use it to differentiate the type of blog the slug is referring to (i.e. integer means it's a dynamic blog loaded from API, NAN means it's a static blog)

## Something worth noticing: 
- Remote authors currently don't have ID. This isn't a problem though since our system doesn't support searching blogs of a specific author.
- Currently only first author will be listed in the post content page.

<img width="600" alt="Blog Index page" src="https://github.com/user-attachments/assets/4351a85d-944f-4858-a5c0-c4a87cb07723">

Below is the correctly rendered remote blog post at `/blog/1`
<img width="400" alt="Rendered Remote Blog Post" src="https://github.com/user-attachments/assets/f38e3023-583e-4703-a9b2-45802146e1e2">

## Other changes: 

- Commit 0188a72f44c509db67b13858e3519360263ebe8e update `BlogTile` UI so only the upper corners of the header image are rounded;
- Commit 21bc201f7266eb934f8fd8d9bbbbbabffc7707b7 fix author alignment with author bio inside blog post (`AuthorTile`), previously it was poorly styled, which made it look like capsule shape sometimes; author placeholder is also added

<img width="659" alt="Screenshot 2024-09-05 at 00 25 19" src="https://github.com/user-attachments/assets/6018cb3a-5aa7-4b23-b5e2-2f1dad934d04">

clementinelove committed:
Update `BlogTile` props to make it more generic

clementinelove committed:
Add author profile image placeholder to `AuthorTile`
- use `faUserCircle` as author profile image placeholder when profile image src is an empty string
- center author profile image to <title + description>: this design doesn't work for very long author bio/description, but works great for current authors.

clementinelove committed:
Merge remote blog as source
- loads blog posts from both: 1. blog API; 2. static markdown pages.

clementinelove commented:
Please have a look at this draft and branch `blog-from-api` when you are free! @JElgar

JElgar commented:
This is awesome thank you! 2 things though:
1. I don't think we need to handle the static blogs. We can just move all of those blogs into the API and then use that directly
2. For the blog id thing, Ive updated the API so the get API uses the slug. Blog writers are already required to provide a unique slug so that should be fine

clementinelove committed:
Remove usage of static posts.

clementinelove committed:
Remove Static Post related functions and update API Client
- Remove static post related functions
- Update API Client to use new API format

clementinelove commented:
API client is now updated accordingly. Static Post related API calls are removed. 
Notice that old markdown post files are not removed, and they are **NOT** added to server yet!

JElgar commented on src/app/blog/utils.ts:
```diff
@@ -15,12 +19,16 @@ export interface Post {
+export async function getStaticPostBySlug(slug: string): Promise<Post> {
   const readFile = fs.readFileSync(`${POSTS_FILE_PATH}/${slug}.md`, "utf-8");
   const {
     data: {
```
Should we remove all of this static stuff now? (And update wording of "remote" stuff to just blog/post

JElgar commented on src/app/blog/utils.ts:
```diff
@@ -48,9 +56,50 @@ export async function getPostBySlug(slug: string): Promise<Post> {
+ * Get a remote blog post using slug.
+ * @param slug URL slug string that denotes a remote blog integer ID.
+ */
+export async function getDynamicPostBySlug(slug: string): Promise<Post | undefined> {
```
And then just expose this directly?

JElgar commented on src/app/blog/[slug]/page.tsx:
```diff
@@ -1,49 +1,53 @@
 }
 
-// TODO Type this properly
-export async function generateStaticParams(): Promise<any> {
```
Im still undecided whether removing this is the right plan. 

The nice thing about it is we can render blogs as soon as they get released on the website. If we keep this then we have to do some sort of build everytime we release a blog which is a bit of a pain.

The annoying thing is
1. We have to fetch the blog each time someone goes to the site (we can solve this with some caching below) but this is a very small issue
2. If we are going to build a site map (which it sounds like we want) then for blogs I assume we will want to know all of them at build time (so we can add all the blogs in the sitemap?)

What do you think makes sense?

clementinelove replied:
I suppose for now we can do it in two ways. One is to fetch everything dynamically (which is basically what this pull request does). I assume you worried about too much API calls right? We can definitely reduce API calls, by:

1. having next.js cache accessed webpages (request results) for a fixed period (see https://nextjs.org/docs/app/api-reference/functions/fetch#optionsnextrevalidate), set it to cache blog posts for a day seems reasonable to me (which is not added to this pull request yet), so that the number of API calls = the number of times Next.js request new data for cache purposes.
2. Same thing applies to sitemaps: we can have next.js dynamically generate a new site map after a fixed period passed.

I think this is the better way to do it. We can do some tests to see if the backend calls reduced after we added dynamic content cache to our project.

As of the second approach: we can also do it fully statically by only fetching data once at build time. I imagine you can add something (called webhooks?) to the backend so that any update to the blogs (or any other part of the backend data) to trigger a new frontend build, which should work just fine if we don't update blogs very frequently. The benefit of this is that we can render the whole website as a static website rather than server-side rendering, in theory because we can distribute static content via CDN the performance would be better. Then again, it will probably take more time to build considering this is an architecture change. 

Let me know if there is anything need to be clarified.

clementinelove committed:
Update API methods with correct query params

clementinelove committed:
Update Blogs Page with a placeholder when failed to fetch data from API

clementinelove committed:
Make `AuthorTile` more generic

clementinelove committed:
Add blog title and subtitle to the blog page.

clementinelove committed:
Remove static related code and posts
